nomad/structs/structs.go

```diff
@@ -7632,6 +7634,39 @@ func (a *Allocation) CopySkipJob() *Allocation {
 	return a.copyImpl(false)
 }
 
+func (a *Allocation) Canonicalize() {
+	if a.AllocatedResources == nil && a.TaskResources != nil {
+		ar := AllocatedResources{}
+		ar.Tasks = toAllocatedResources(a.TaskResources)
+
+		if a.SharedResources != nil {
```

Comment:
`a.AllocatedResources == nil && a.TaskResources == nil` is an invalid state - as it would mean either than the alloc doesn't have tasks or isn't assigned any resources and alloc runner should reject it.  It's tested in `TestClient_AddAllocError ` test: https://github.com/hashicorp/nomad/blob/v0.10.0/client/client_test.go#L760-L762 .

Current master state: Allocs created Pre-0.9, TaskResources should always be non-nil; Allocs created post-0.9, TaskResources and AllocatedResources should both be non-nil.